000-orm-in-databuilder.md

```diff
@@ -0,0 +1,97 @@
+    - usage(entity_type, entity_key, user_email, read_count)
+    
+    - dashboard(dashboard_key, dashboard_name, dashboard_created_timestamp, dashboard_url, dashboard_description_key, dashboard_description, last_execution_key, last_execution_timestamp, last_execution_state, last_modified_key, last_modified_timestamp, last_modified_timestamp_name, group_key, group_name, group_url, group_description_key, group_description, cluster_key, cluster_name)
+    
+    - dashboard_query(query_key, query_id, query_name, query_url, query_text, dashboard_key)
+    
+    - dashboard_chart(chart_key, chart_id, chart_name, chart_type, chart_url, query_key)
+    
+    - dashboard_table(dashboard_key, table_key)
+    
+    - badge(badge_key, badge_category, entity_key)
+    
+- Due to the reuse of graph db models, add new transformers for entities to convert 'node' and 'relationship' data to 'record' with model translators. 
+
+- Add a new data loader(e.g., file_system_mysql_csv_loader) to consume transformed metadata records and dump them into csv to a new 'records' folder. The new data loader works with the ORM model translator and hence, the csv output could be consistent with the schema we defined for ORM.
+
+- Add a new publisher(e.g., mysql_csv_publisher) to publish the loaded csv files from 'records' folder to MySQL. The new publisher works with the ORM model translator to create schema if necessary and is responsible for ORM CRUD operations.
+
+- Add a new elastic search publisher supporting ORM operations
+
+## Drawbacks
+
+- Support of ORM would get a big change involved in model translator, data loader, transformer and publisher. 
+Transformer for model translator might downgrade the performance due to the conversion from node/rel to record.
+
+- We can still use all the existing graph db models, though need to keep the ORM model translator compatible with future change in graph db model. 
+
+- Even if with a schema upgrade feature, relational db is not as flexible as graph db to handle metadata structure change
+## Alternatives
+
+- Instead of using model translators, develop complete ORM models containing schema definition and record generator(e.g., `create_record_iterator()`), though it will lead to two model versions for each entity. On the other hand, we may have to duplicate all existing extractors and make them work with the new ORM models.
+Moreover, we will have to ensure any future changes compatible not only in two models but also in two extractors.
+
+- Use normalized models to design schema in relational DBs. The disadvantage is it will hurt query performance due to multiple joins.
+
+## Prior art
+
+N/A
+## Unresolved questions
+
+- The initial relational DB schema plan was made according to the current graph DB models and it may need some polishment from Amundsen staff who are more familiar with real use cases.
+- Regarding how to map graph db models to ORM models, it would be great if there is a dynamic way for the model conversion
```

Comment:
+1, but aside from performance, some of the metadata can't be fetched/gathered in single extraction as they could come from different sources.

I am mostly curious on how to get all those info into single table row with just one extraction.